### PR TITLE
Dev qa

### PR DESCRIPTION
**Pull request by vaslee**
merge all the code we had into master, these codes should be the newest code

### CODE REVIEW

File: StellarApp/Controllers/SolarViewController.swift [hunk 1]
```diff
@@ -1,75 +1,189 @@
...
+//            sceneView.removeGestureRecognizer(tapGestureRecognizer)
+//            sceneView.removeGestureRecognizer(pinchGestureRecognizer)
+//            sceneView.scene.rootNode.enumerateChildNodes { (node, stop ) in
+//                node.removeFromParentNode()
+//            }
+//            sceneView.scene.rootNode.addChildNode(centerNode)
+//            MovedPlanet.getPlanets().forEach { centerNode.addChildNode($0) }
+//            sceneView.isUserInteractionEnabled = false
+//
+//        }
```
Comment (vhart) [inline StellarApp/Controllers/SolarViewController.swift]: commented out code needed?

File: StellarApp/Controllers/SolarViewController.swift [hunk 2]
```diff
@@ -1,75 +1,189 @@
...
     }
     
+   
+    
+    private func setUpSolarView() {
+        solarView.translatesAutoresizingMaskIntoConstraints = false
+
+        sceneView.addSubview(solarView)
+        sceneView.bringSubviewToFront(solarView)
```
Comment (vhart) [inline StellarApp/Controllers/SolarViewController.swift:94]: is this needed?

File: StellarApp/Controllers/SolarViewController.swift [hunk 3]
```diff
@@ -1,75 +1,189 @@
...
+        vc.modalTransitionStyle = .crossDissolve
+        self.present(vc, animated: true, completion: nil)
+    }
+    
+    @objc func pinch(pinch: UIPinchGestureRecognizer) {
+        let pinchView = pinch.view as! SCNView
+        let pinchLocation = pinch.location(in: pinchView)
+        let hitTest = pinchView.hitTest(pinchLocation, options: nil)
+        if !hitTest.isEmpty {
+            //            let scaleAction = SCNAction.scale(by: pinch.scale, duration: 0)
```
Comment (vhart) [inline StellarApp/Controllers/SolarViewController.swift]: needed?

File: StellarApp/Models/NewsDataManager.swift [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+                    print("property list is encoding error: \(error)")
+                }
+            } else {
+                print("data is nil")
+            }
+        }
+        return news
+    }
+
+    static func delete(news: NewsData, atIndex index: Int) {
```
Comment (vhart) [inline StellarApp/Models/NewsDataManager.swift]: Is anything in here needed? Is this file needed?

File: StellarApp/Models/PlanetsInfo.swift [hunk 5]
```diff
@@ -1,9 +1,112 @@
...
+//        self.numberOfMoons = dict[PlanetCollectionKeys.NumberOfMoonsKey] as? String ?? "0"
+//        self.description = dict[PlanetCollectionKeys.DescriptionKey] as? String ?? ""
 //
-//  PlanetsInfo.swift
-//  StellarApp
-//
-//  Created by TingxinLi on 4/8/19.
-//  Copyright © 2019 TingxinLi. All rights reserved.
-//
+//    }
```
Comment (vhart) [inline StellarApp/Models/PlanetsInfo.swift]: delete this

File: StellarApp/Models/PlanetsInfo.swift [hunk 6]
```diff
@@ -1,9 +1,112 @@
...
+    lazy var planetMapping: [PlanetType: PlanetInfo] = {
+        let planetData = getPlanetData()
+        var planetDict: [PlanetType: PlanetInfo] = [:]
+        for info in planetData {
+            if let planetType = PlanetType(rawValue: info.name.lowercased()) {
+                planetDict[planetType] = info
+            }
+//                else {
+//                fatalError("unhandled planet type \(dump(info))")
+//            }
```
Comment (vhart) [inline StellarApp/Models/PlanetsInfo.swift:56]: Make a Trello Card to come back to this. Essential this is saying that you guys have planets that are not being handled.

File: StellarApp/Protocol/CubeMapWall.swift [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        // add masked box
+//        let maskedBox = SCNBox(width: thickness, height: height, length: length, chamferRadius: 0)
+//        maskedBox.firstMaterial?.diffuse.contents = UIColor.white
+//        
+//        maskedBox.firstMaterial?.transparency = 0.00001
+//        
+//        let maskedBoxNode = SCNNode(geometry: maskedBox)
+//        maskedBoxNode.renderingOrder = 100
+//        maskedBoxNode.position = SCNVector3.init(thickness, 0, 0)
+//        self.addChildNode(maskedBoxNode)
```
Comment (vhart) [inline StellarApp/Protocol/CubeMapWall.swift]: Since we removed the mask, can we delete this? You can try re-adding it first. What I noticed was that the mask was glitching through the wall on one face of the box, perhaps its position was off?

File: StellarApp/Protocol/Planet.swift [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+    
+    func stopRotating() {
+        self.removeAction(forKey: "rotateNode")
+    }
+    
+}
+
+
+
+
```
Comment (vhart) [inline StellarApp/Protocol/Planet.swift]: can we split this file up?

File: StellarApp/Protocol/Planet.swift [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+struct RotationVector {
+    let x: CGFloat, y: CGFloat, z: CGFloat
+}
+
+protocol Rotatable {
+    func rotate()
+    func stopRotating()
+}
+
+class RotatingNode: SCNNode, Rotatable {
```
Comment (vhart) [inline StellarApp/Protocol/Planet.swift]: you never made use of this class. This is a possible way to refactor but for now it's not being used so I think you should delete it

File: StellarApp/Protocol/Planet.swift [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+
+extension SCNNode {
+    static func deepScaleNode(node: SCNNode, scaleAction: SCNAction) {
+        node.runAction(scaleAction)
+        for child in node.childNodes {
+            deepScaleNode(node: child, scaleAction: scaleAction)
+        }
+    }
+    
+    static func deepScaleNode(node: SCNNode, scale: CGFloat) {
```
Comment (vhart) [inline StellarApp/Protocol/Planet.swift]: we using both of these `deepScaleNode` functions? If not delete the unused one.

File: StellarApp/Protocol/Planet.swift [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+        let newScale = SCNVector3Make(Float(scaleX), Float(scaleY), Float(scaleZ))
+        node.scale = newScale
+        for child in node.childNodes {
+            deepScaleNode(node: child, scale: scale)
+        }
+    }
+    
+    
+    
+    static func getNode(name: String, image: UIImage,
```
Comment (vhart) [inline StellarApp/Protocol/Planet.swift:138]: to make this closer to what the others indicate `getRing` etc. Maybe call this `getSphere`

File: StellarApp/Protocol/Planet.swift [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+        
+        ringGeo = SCNTorus(ringRadius: radius, pipeRadius: 0.01)
+        ringNode.geometry = ringGeo
+        ringNode.position = SCNVector3(x: vector.x, y: vector.y, z: vector.z)
+        ringGeo.firstMaterial?.diffuse.contents = color
+        
+        return ringNode
+    }
+    
+    static func getSaturnRing(radius: CGFloat,
```
Comment (vhart) [inline StellarApp/Protocol/Planet.swift:170]: this is the exact same function as the one above it. Get rid of this one and use the more generic one

File: StellarApp/Protocol/PlanetsMoved.swift [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+                planetNodes.append(orbitNode)
+            }
+        }
+        return planetNodes
+    }
+}
+
+extension SCNNode {
+    static func deepScaleNodes(node: SCNNode, scaleAction: SCNAction) {
+        node.runAction(scaleAction)
```
Comment (vhart) [inline StellarApp/Protocol/PlanetsMoved.swift]: why is this extension in two files. Its the same function. Extensions should be in a folder called `Extensions` and the File name should be class/struct/obj + extension purpose, ex: `SCNNode+NodeBuilding.swift
`

File: StellarApp/Protocol/PlanetsMoved.swift [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+        ringGeo.firstMaterial?.diffuse.contents = texture
+        
+        
+        return ringNode
+    }
+    
+}
+
+struct Sound {
+static var player: AVAudioPlayer?
```
Comment (vhart) [inline StellarApp/Protocol/PlanetsMoved.swift]: move to its own file

File: StellarApp/Protocol/PlanetsWithAction.swift [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+                planetNodes.append(orbitNode)
+                
+            default:
+                break
+            }
+        }
+        return planetNodes
+    }
+    
+
```
Comment (vhart) [inline StellarApp/Protocol/PlanetsWithAction.swift:117]: extra lines can be deleted

File: StellarApp/Protocol/PlanetsWithAction.swift [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+            }
+        }
+        return planetNodes
+    }
+    
+    
+}
+
+extension SCNNode {
+    static func deepperScaleNode(node: SCNNode, scaleAction: SCNAction) {
```
Comment (vhart) [inline StellarApp/Protocol/PlanetsWithAction.swift:121]: another set of extension functions on SCNNode. Figure out which ones are being used, delete the rest.

File: StellarApp/Controllers/SolarViewController.swift (hunk 1 above)
Comment (vaslee) [reply StellarApp/Controllers/SolarViewController.swift]: yes, its for the play button, before we click on it, all the planets are not moving, after we click it, all the planets will be moving around the sun

Comment (vaslee) [reply StellarApp/Controllers/SolarViewController.swift]: but since last time you help me with code, we haven't talked about how to refactor these codes.

File: StellarApp/Models/NewsDataManager.swift (hunk 4 above)
Comment (vaslee) [reply StellarApp/Models/NewsDataManager.swift]: I will check it out

File: StellarApp/Models/PlanetsInfo.swift (hunk 6 above)
Comment (vaslee) [reply StellarApp/Models/PlanetsInfo.swift:56]: we need it for now.

File: StellarApp/Protocol/CubeMapWall.swift (hunk 7 above)
Comment (vaslee) [reply StellarApp/Protocol/CubeMapWall.swift]: I got.

File: StellarApp/Protocol/Planet.swift (hunk 12 above)
Comment (vaslee) [reply StellarApp/Protocol/Planet.swift:170]: they are different because for Saturn orbit we had a texture for it instead of the color.

File: StellarApp/Protocol/PlanetsMoved.swift (hunk 13 above)
Comment (vaslee) [reply StellarApp/Protocol/PlanetsMoved.swift]: ok, this is something I want to tell you about regarding planets' orbits rotation. can we work on this together when we meet tomorrow?

File: StellarApp/Protocol/PlanetsWithAction.swift (hunk 16 above)
Comment (vaslee) [reply StellarApp/Protocol/PlanetsWithAction.swift:121]: The same problem.. we will talk about it tomorrow

File: StellarApp/Protocol/Planet.swift (hunk 9 above)
Comment (vaslee) [reply StellarApp/Protocol/Planet.swift]: we need it... we talk about this tomorrow, its relating to the code we need to work on.

File: StellarApp/Controllers/SolarViewController.swift [hunk 17]
```diff
@@ -1,75 +1,189 @@
...
+            SCNNode.deepScaleNode(node: centerNode, scale: pinch.scale)
```
Comment (vhart) [inline StellarApp/Controllers/SolarViewController.swift]: need to make sure not to scale the cube map node

File: StellarApp/Models/NewsDataManager.swift (hunk 4 above)
Comment (vhart) [reply StellarApp/Models/NewsDataManager.swift]: Deleting?

File: StellarApp/Controllers/NewsViewController.swift [hunk 18]
```diff
@@ -1,30 +1,93 @@
...
+    
+    let newsView = NewsView()
     override func viewDidLoad() {
         super.viewDidLoad()
-
-        // Do any additional setup after loading the view.
+        view.addSubview(newsView)
+        newsView.newsCollectionView.delegate = self
+        newsView.newsCollectionView.dataSource = self
+        getArticles(keyword: "astronomy")
```
Comment (vhart) [inline StellarApp/Controllers/NewsViewController.swift]: I think you guys can include multiple keywords

File: StellarApp/Protocol/Planet.swift (hunk 8 above)
Comment (vhart) [reply StellarApp/Protocol/Planet.swift]: Was this resolved?

File: StellarApp/Protocol/Planet.swift (hunk 12 above)
Comment (vhart) [reply StellarApp/Protocol/Planet.swift:170]: ok sounds good. Then maybe just remove the `Saturn` part. `getRing`

**Review by vhart — COMMENTED** [review]
Still a few outstanding questions. If you need to merge though, I won't block you, just let me know

**Comment by vaslee** [comment]
new update codes
